File: Python/221. Maximal Square.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        rows = len(matrix)
        cols = 0 if rows == 0 else len(matrix[0])
        dp = [[0 for j in range(cols + 1)] for i in range(rows + 1)]
        maxlen = 0        
        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                if matrix[i - 1][j - 1] == '1':                    
                    dp[i][j] = min(dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1]) + 1
                    maxlen = max(maxlen, dp[i][j])
        return maxlen * maxlen

File: 2024/221. Maximal Square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        vector<vector<int>> dp(matrix.size(), vector<int>(matrix[0].size(), 0));
        int res = 0;
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                if (matrix[i][j] == '1') {
                    if (i > 0 && j > 0) {
                        dp[i][j] = min(dp[i - 1][j],
                                       min(dp[i][j - 1], dp[i - 1][j - 1]));
                    }
                    dp[i][j] += 1;
                }
                res = max(dp[i][j], res);
            }
        }
        return res * res;
    }
};